File: 0216-combination-sum-iii/0216-combination-sum-iii.swift
class Solution {
    func combinationSum3(_ k: Int, _ n: Int) -> [[Int]] {
        func backtrack(_ total: Int, _ curr: inout [Int]) {
            if curr.count == k {
                if total == n {
                    res.append(curr)
                }
                return
            } else if total >= n {
                return
            }
            for num in (curr.last ?? 0) + 1 ..< 10 {
                curr.append(num)
                backtrack(total + num, &curr)
                curr.removeLast()
            }
        }

        var res: [[Int]] = []
        var tempCurr: [Int] = []
        backtrack(0, &tempCurr)
        return res
    }
}

File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> [[int]]:
        def backtrack(total: int, curr: [int]):
            if len(curr) == k:
                if total == n:
                    res.append(curr.copy())
                return
            elif total >= n:
                return
            for num in range(curr[-1] + 1 if curr else 1, 10):
                curr.append(num)
                backtrack(total + num, curr)
                curr.pop()


        res = []
        tempCurr = []
        backtrack(0, tempCurr)
        return res
